### ISSUE TITLE
Figure out a way to manage the discrepancy on windows nodes from the linux node

### ISSUE THREAD

**Issue by dchen1107** [kubernetes/kubernetes#60338]
https://github.com/kubernetes/kubernetes/pull/60275/files/2d942dab68b64e684dd2b75232cf9daabc6e0a95#diff-10055ae93a8699af13ceba0482fc43c3

I am expecting we have more windows specific config like code coming to Kubelet and other node components. We should figure out a way to handle this, so that the code can be easily managed? Instead of letting windows-special handling code like this scattered everywhere in our codebase. 

cc/ @feiskyer @michmike  @kubernetes/sig-windows-bugs

**Comment by feiskyer** [comment]
@dchen1107 Thanks. This is indeed a problem of windows containers. Usually, we handle this problem in two ways

- (1) By checking `runtime.GOOS == "windows"`
- (2) By splitting codes into platform specific files, e.g. `file_linux.go`, `file_windows.go`, `file_unsupported.go`

I think (2) is our preferred way,  but it is also broken frequently. #51540 is aiming to add windows node e2e tests, but the testing environment is still not ready yet. I'd like to clean up related codes after that.

**Comment by yujuhong** [comment]
@feiskyer is there anything left to do for this bug?

**Comment by feiskyer** [comment]
@yujuhong this is actually cleanup instead of bug. I think there're still many places needing to improve.

/kind cleanup

**Comment by pacoxu** [comment]
> (2) By splitting codes into platform specific files, e.g. file_linux.go, file_windows.go, file_unsupported.go

(2) is current preferred way.  For `cmd/kubelet` and `pkg\kubelet`, there are 40 matches across 17 files for `runtime.GOOS`.

/triage accepted
/help
/good-first-issue

**Comment by pmgk07** [comment]
I'd like to help out on this issue. Needed pointers on if anything more than the below is required:
- I find a file with runtime.GOOS 
- Move that piece of code in "file" to "file_linux" / "file_windows" / "file_others"
- Tag the platform specific things with // +build linux ,  // +build windows

**Comment by adisky** [comment]
> I'd like to help out on this issue. Needed pointers on if anything more than the below is required:
> 
> * I find a file with runtime.GOOS
> * Move that piece of code in "file" to "file_linux" / "file_windows" / "file_others"
> * Tag the platform specific things with // +build linux ,  // +build windows

@pmgk07 right, this has to be done, Make sure do it for all runtime.GOOS(wherever possible) instances

**Comment by pmgk07** [comment]
Thanks @adisky. Raising a simple PR for ./cmd/kubelet/options package to just get the idea of a general PR workflow in this repo

**Comment by sorkinl** [comment]
Is this still in need of cleanup? If not I'd like to work on this. @pmgk07 
/assign

**Comment by pmgk07** [comment]
@sorkinl Yes, `pkg/kubelet` contains runtime.GOOS checks within them.
```
$ grep -rn "runtime.GOOS" pkg/kubelet|wc -l 
      61
```

**Comment by sorkinl** [comment]
> > I'd like to help out on this issue. Needed pointers on if anything more than the below is required:
> > 
> > * I find a file with runtime.GOOS
> > * Move that piece of code in "file" to "file_linux" / "file_windows" / "file_others"
> > * Tag the platform specific things with // +build linux ,  // +build windows
> 
> @pmgk07 right, this has to be done, Make sure do it for all runtime.GOOS(wherever possible) instances

Thank you @pmgk07 ! I will try to follow these instructions and post here with updates.

**Comment by raphminkyu** [comment]
Hi @pmgk07, I'm working on the issue with @sorkinl . We tried making some changes and I was wondering if you could take a look and comment:

We made changes in pkg/kubelet/volume_host.go and changed 
`func (kvh *kubeletVolumeHost) GetPodVolumeDir(podUID types.UID, pluginName string, volumeName string) string {
	dir := kvh.kubelet.getPodVolumeDir(podUID, pluginName, volumeName)
	if runtime.GOOS == "windows" {
		dir = util.GetWindowsPath(dir)
	}
	return dir
}`

and made "pkg/kubelet/volume_host_windows.go":
`//go:build windows
// +build windows

package kubelet

func (kvh *kubeletVolumeHost) GetPodVolumeDir(podUID types.UID, pluginName string, volumeName string) string {
	dir := util.GetWindowsPath(dir)
	return dir
}`
Is this the right approach? Do we test with `go test ./pkg/kubelet/...`?

**Comment by pmgk07** [comment]
@raphminkyu Sorry for the late reply on this PR. Yes, that looks good enough.

So far these are the files that are in need of refactoring. I grepped on "runtime.GOOS" occurences, might have some false positives.

- pkg/kubelet/cm/devicemanager/manager.go
- pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
- pkg/kubelet/kubelet_test.go
- pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go
- pkg/kubelet/stats/cri_stats_provider_test.go
- pkg/kubelet/kuberuntime/kuberuntime_container.go
- pkg/kubelet/lifecycle/predicate_test.go
- pkg/kubelet/lifecycle/predicate.go
- pkg/kubelet/kubelet_pods.go
- pkg/kubelet/nodestatus/setters.go
- pkg/kubelet/volume_host.go
- pkg/kubelet/kubelet.go
- pkg/kubelet/kubelet_node_status_test.go
- pkg/kubelet/kubelet_node_status.go

@raphminkyu & @sorkinl I'll pick up kubelet_pods / kubelet / kubelet_node_status files (& their corresponding test files) for refactoring.

**Comment by sorkinl** [comment]
@pmgk07 Thank you this helps a lot. We will start with changes going down the list and will reach out if we have any questions.

**Comment by yujuhong** [comment]
@pacoxu or others, do you know if we made much progress on this?

**Comment by pacoxu** [comment]
> 2022 status
> * pkg/kubelet/cm/devicemanager/manager.go
> * pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
> * pkg/kubelet/kubelet_test.go
> * pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go
> * pkg/kubelet/stats/cri_stats_provider_test.go
> * pkg/kubelet/kuberuntime/kuberuntime_container.go
> * pkg/kubelet/lifecycle/predicate_test.go
> * pkg/kubelet/lifecycle/predicate.go
> * pkg/kubelet/kubelet_pods.go
> * pkg/kubelet/nodestatus/setters.go
> * pkg/kubelet/volume_host.go
> * pkg/kubelet/kubelet.go
> * pkg/kubelet/kubelet_node_status_test.go
> * pkg/kubelet/kubelet_node_status.go

Current status for pkg/kubelet. No `runtime.GOOS` in `cmd/kubelet`

- pkg/kubelet/kubelet_pods.go
- pkg/kubelet/kubelet_server_journal.go
- pkg/kubelet/volume_host.go
- pkg/kubelet/cm/devicemanager/manager.go
- pkg/kubelet/cm/memorymanager/memory_manager.go
- pkg/kubelet/eviction/eviction_manager.go
- pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
- pkg/kubelet/lifecycle/predicate.go

TODO items, and a quick search, there is no WIP PR now if I do not miss some cleanup PRs.

**Comment by serjLozovskoy** [comment]
Is it still open? May I work with it issue?

**Comment by BHAVISHYA2005** [comment]
I think it's better to close the issue, No updates for so long 🙂
